include/caffe/solver.hpp

```diff
@@ -159,6 +158,22 @@ class RMSPropSolver : public SGDSolver<Dtype> {
 };
 
 template <typename Dtype>
+class AdaDeltaSolver : public SGDSolver<Dtype> {
+ public:
+  explicit AdaDeltaSolver(const SolverParameter& param)
+      : SGDSolver<Dtype>(param) { PreSolve(); }
+  explicit AdaDeltaSolver(const string& param_file)
+      : SGDSolver<Dtype>(param_file) { PreSolve(); }
+
+ protected:
+  void PreSolve();
```

Comment:
I think it is better to rename `AdaDeltaSolver::PreSolve()` into `AdaDeltaSolver::AdaDeltaPreSolve()`. Since you are going to call AdaDeltaSolver's presolve function after SGDSolver's presolve function, it is better to avoid a name conflict with `SGDSolver::PreSolve()`.